keps/sig-storage/2317-fsgroup-on-mount/README.md

```diff
@@ -0,0 +1,400 @@
+may not support them. This enhancement proposes providing the CSI driver with fsgroup as an explicit field so as CSI driver can apply this on mount time.
+
+## Motivation
+
+Since some CSI drivers(AzureFile for example) - don't support chmod/chown - we propose that fsgroup of pod to be provided to the CSI driver on `NodeStage`
+and `NodePublish` CSI RPC calls. This will allow the CSI driver to apply `fsgroup` as a mount option on `NodeStage` or `NodePublish` and kubelet can be freed
+from responsibility of applying recursive ownership and permission change.
+
+This feature hence becomes a prerequisite of CSI migration of Azure file driver and removal of Azure Cloud Provider.
+
+### Goals
+
+- Allow CSI driver to mount volumes with provided fsgroup.
+
+### Non-Goals
+
+- We are not supplying `fsGroup` as a generic ownerhip and permission handle to the CSI driver. We do not expect CSI drivers to `chown` or `chmod` files.
+
+## Proposal
+
+We are updating CSI specs by adding additional field called `volume_mount_group` to `NodeStage` and `NodePublish` RPC calls. The CSI proposal is available at - https://github.com/container-storage-interface/spec/pull/468 .
+
+The CSI spec change is deliberately trying to avoid asking drivers to use supplied `fsGroup` as a generic handle for ownership and permissions. The reason being - Kubernetes may expect ownership and permissions to be in a way that is very platform/OS specific. We do not think CSI driver is right place to enforce all kind of different permissions expected by Kubernetes. The full scope of that discussion is out of scope for this enhancement and interested folks can follow along on - https://github.com/container-storage-interface/spec/issues/449
+
+
+### Risks and Mitigations
+
+I am not aware of any associated risks. If a driver can not support using `fsgroup` as a mount option, it can always use `FileFSGroupPolicy` and let kubelet handle the ownership and permissions.
+
+## Design Details
+
+We are proposing that when kubelet determines a CSI driver has `VOLUME_MOUNT_GROUP` node capability, the kubelet will use proposed CSI field `volume_mount_group` to pass pod's `fsGroup` to the CSI driver. Kubelet will expect that driver will use
+this field for mounting volume with given `fsGroup` and no further permission/ownerhip change will be necessary.
+
+It should be noted that if a CSI driver advertises `VOLUME_MOUNT_GROUP` node capability then value defined in `CSIDriver.Spec.FSGroupPolicy` will be ignored and kubelet willl always use `fsGroup` as a mount option.
+
+### Test Plan
+
+Unit test:
+1. Test that whenever supported pod's `fsGroup` should be passed to CSI driver via `volume_mount_group` field.
+
+For alpha feature:
+1. Update Azure File CSI driver to support supplying `fsGroup` via `NodeStage` and `NodePublish`.
```

Comment:
There should be tests for non-Azure File CSI driver to make sure there are no regressions.

Reply:
There are already tests that verify fsgroup behaviour for existing drivers.  I do not think we need specifically new tests for them - I think.